1st radio-api

AlexErk opened the pull request:
1st draft of the radio api

Commit:
1st radio-api

1st draft of the radio api

AlexErk commented:
Dear all this is a 1st version in order to integrate teh Radio use cases into the current draft of the API. In general I've done the following:

1: Changed the names of all the interfaces, which are applicable for Radio and TV.

2: Introduced a ChannelAttributes interface and sub interfaces to handle technology specific channel attributes. IHMO this is needed for the different TV systems too.

3: Introduced a Application interface. This Application interface could handle the differen user applications, TV and Radio services might have. 

Open Questions: 
1: TVMediaStream should we rename it ? 

2: There's a oneitbroadcasted EventHandler in Source why? I would assume, that when EIT data is available, it's handled via the getPrograms method in the Channel interface.

3: TVCICards is the last only TV named interface just because there are no CI cards in Radio

tidoust commented:
Hello @AlexErk,

@stevem-tw kindly agreed to edit the spec. To ensure he start with "interesting" editing tasks as soon as possible, I offered to convert the spec first to the so-called WebIDL contiguous syntax (see issue #6) that ReSpec now requires.

It would be good to merge this Pull Request before I run the conversion though, as it touches on different parts of the spec. I think the main blocker at this stage is the absence of prefix for interfaces (issue #16). While the group discusses possible prefixes, could you rollback on your changes and re-introduce the `TV` prefix everywhere?

I believe we can merge the PR afterwards and continue discussing possible updates to these interfaces within the group and in subsequent PRs.

chrisn commented on index.html:
```diff
@@ -1500,120 +1491,243 @@
+    <section>
+      <h2><a>DABChannelAttributes</a> Interface</h2>
+      <p>
+        The <a>DABChannelAttributes</a> interface represents DVB channel 
```
DAB

chrisn commented on index.html:
```diff
@@ -1500,120 +1491,243 @@
+
+        <dt>readonly attribute DOMString serviceId</dt>
+          <dd>
+            MUST return the identification of aDAB radio or data service
```
"a DAB"

tidoust committed:
Re-introduced "TV" prefix on all interfaces

Interfaces needs a prefix for namespacing not to confuse developers and avoid
possible name collisions with other specs. Note the group may decide to use
another prefix than "TV" in the end.

The update also adds missing inheritance to EventTarget for the interfaces
that were recently introduced and that use events.

tidoust commented:
@AlexErk I created a PR on your PR that re-introduces the `TV` prefix and leaves the rest untouched:
https://github.com/AlexErk/tvcontrol-api/pull/1

If you can review and merge that PR on your fork, the update should appear here automatically.

chrisn commented on index.html:
```diff
@@ -1653,78 +1767,78 @@
+            MUST return the associated  program of the TV/Radio recording if applicable.
             Please note the recording should keep track of the relevant updates
-            in Event Information Table (EIT) while <a>program</a> is specified,
+            in Event Information Table (EIT) inthe case of DVB or ProgramInfo (in the case of DAB) while <a>Program</a> is specified,
```
"in the" and "program"

chrisn commented on index.html:
```diff
@@ -2349,160 +2463,208 @@
 
       <p>
-        The attribute <code>operation</code> in the <a>TVTunerChangedEvent</a>
+        The attribute <code>ApplicationType</code> in a <a>Application</a> can have the
```
"type"

chrisn commented on index.html:
```diff
@@ -1500,120 +1491,243 @@
+          attribute EventHandler onapplicationchanged
+        </dt>
+          <dd>
+            Handles the <code>onapplicationchanged</code> event, fired when
```
Is this event handler triggered when the `applicationData` changes?

chrisn reviewed (commented):
Your note in the PR asks a good question:

> Should we introduce more specific application type interfaces here. e.g. for the DAB/DVB User applications (Slideshow, DL, Journaline, HbbTV, MHEG)? Or are we happy with just delivering the applicationData to the developers.

I think my own preference would be to have application-specific interfaces, e.g, a Slideshow interface such as

```
interface Slideshow : Application {
  attribute EventHandler onimagereceived;
};
```

which would give us a place to put other methods, if needed, and allow the attributes to more meaningful names than the generic `applicationData`.

WDYT, @tidoust?

tidoust commented:
@chrisn asked:

> Should we introduce more specific application type interfaces here. e.g. for the DAB/DVB User applications (Slideshow, DL, Journaline, HbbTV, MHEG)? Or are we happy with just delivering the applicationData to the developers.

That seems like a place in the spec where we would like others (or other specs) to be able to define extensions. A similar example that comes to mind is the work on the [Generic Sensor API](https://w3c.github.io/sensors/). From an IDL perspective, it defines a base `Sensor` interface and let other specs define concrete sensors interface that inherit from `Sensor`. It seems a good idea to follow the same path here.

> I think my own preference would be to have application-specific interfaces

I agree. We probably do not want a fixed enumeration for application types in any case.

chrisn commented:
A few more comments on these changes:

Is application data only available for channels being presented, or can the data from any channel be accessed? I'm wondering whether the application data should be more closely related to the `TVMediaStream` rather than the `TVChannel` interface.

Is the application data expected to be aligned temporally with the audio/video media? If so, we could look at using [TextTrackCue](https://w3c.github.io/html/single-page.html#texttrackcue-texttrackcue), as `TVMediaStream` has support for text tracks. Maybe we could use [DataCue](https://w3c.github.io/html/single-page.html#datacue-datacue) to support application-specific data types.

I'm also worried that introducing an `Application` (or `TVApplication`) interface might be confusing, as elsewhere in the spec we use "application" to mean Web applications that use the TV Control API.

Having said this, I think we should merge this PR (together with Francois' changes) and raise separate issues for these questions.

AlexErk commented:
1: I would assume, that application data is only to be presented for the channel they belong to.  

2: Yes application datte (Slideshow and textmessges ) are aligened temporall with the audio. I would alos assume that for many other types is will be very likely. So the TextTrackCue might fit perfect for text messages. 

3:  I agree Application might be misleading. Suggestion "DataComponent"

tidoust commented:
Hi @AlexErk,

Note that, to provide @stevem-tw with a spec in the "right" WebIDL format, as noted in a previous comment [1], I will merge my PR (to re-introduce the `TV` prefix) and your PR and run the conversion early next week. I would expect the discussion to continue and interfaces to evolve afterwards, of course!

[1] https://github.com/w3c/tvcontrol-api/pull/14#issuecomment-255354089

Commit:
Merge pull request #1 from tidoust/radio-api

Re-introduced "TV" prefix on all interfaces
